File: kore-node/kore_fileformat.test.ts
/**
 * KORE Node.js FFI Integration Tests
 * ===================================
 *
 * Tests for Node.js N-API wrapper around Rust kore-ffi library.
 */

import * as kore from './kore_fileformat';
import * as fs from 'fs';
import * as path from 'path';
import * as assert from 'assert';

// ─────────────────────────────────────────────────────────────────────────────
// DATA TYPES
// ─────────────────────────────────────────────────────────────────────────────

describe('DataTypes', () => {
  it('should have correct DataType enum values', () => {
    assert.strictEqual(kore.DataType.I64, 1);
    assert.strictEqual(kore.DataType.F64, 2);
    assert.strictEqual(kore.DataType.BOOL, 3);
    assert.strictEqual(kore.DataType.STR, 4);
    assert.strictEqual(kore.DataType.STR_DICT, 5);
    assert.strictEqual(kore.DataType.ARRAY, 6);
    assert.strictEqual(kore.DataType.STRUCT, 7);
  });

  it('should have correct Compression enum values', () => {
    assert.strictEqual(kore.Compression.RAW, 0);
    assert.strictEqual(kore.Compression.RLE, 1);
    assert.strictEqual(kore.Compression.DELTA, 2);
    assert.strictEqual(kore.Compression.DICT, 3);
    assert.strictEqual(kore.Compression.NAN_RAW, 4);
    assert.strictEqual(kore.Compression.DEFLATE, 5);
    assert.strictEqual(kore.Compression.ZSTD, 6);
  });
});

// ─────────────────────────────────────────────────────────────────────────────
// DATA BLOCK
// ─────────────────────────────────────────────────────────────────────────────

describe('DataBlock', () => {
  it('should create empty data block', () => {
    const block = kore.createDataBlock();
    assert.strictEqual(block.numRows, 0);
    assert.strictEqual(block.numColumns, 0);
    assert.strictEqual(block.columns.length, 0);
  });

  it('should add single column', () => {
    const block = kore.createDataBlock();
    block.addColumn('numbers', kore.DataType.I64, [1, 2, 3, 4, 5]);

    assert.strictEqual(block.numRows, 5);
    assert.strictEqual(block.numColumns, 1);
    assert(block.getColumn('numbers') !== undefined);
  });

  it('should add multiple columns', () => {
    const block = kore.createDataBlock();
    block.addColumn('numbers', kore.DataType.I64, [1, 2, 3]);
    block.addColumn('names', kore.DataType.STR, ['a', 'b', 'c']);

    assert.strictEqual(block.numRows, 3);
    assert.strictEqual(block.numColumns, 2);
  });

  it('should reject mismatched row counts', () => {
    const block = kore.createDataBlock();
    block.addColumn('numbers', kore.DataType.I64, [1, 2, 3]);

    assert.throws(() => {
      block.addColumn('names', kore.DataType.STR, ['a', 'b']); // 2 rows != 3
    });
  });

  it('should retrieve column by name', () => {
    const block = kore.createDataBlock();
    block.addColumn('test', kore.DataType.I64, [10, 20, 30]);

    const col = block.getColumn('test');
    assert(col !== undefined);
    assert.strictEqual(col!.name, 'test');
    assert.strictEqual(col!.dtype, kore.DataType.I64);
    assert.deepStrictEqual(col!.data, [10, 20, 30]);
  });

  it('should return undefined for non-existent column', () => {
    const block = kore.createDataBlock();
    block.addColumn('test', kore.DataType.I64, [1, 2, 3]);

    assert.strictEqual(block.getColumn('nonexistent'), undefined);
  });
});

// ─────────────────────────────────────────────────────────────────────────────
// ROUNDTRIP (Phase 3 placeholder - JSON fallback)
// ─────────────────────────────────────────────────────────────────────────────

describe('Roundtrip', () => {
  it('should write and read data block', async () => {
    const block = kore.createDataBlock();
    block.addColumn('numbers', kore.DataType.I64, [1, 2, 3, 4, 5]);
    block.addColumn('names', kore.DataType.STR, ['a', 'b', 'c', 'd', 'e']);

    const tmpFile = path.join('/tmp', `kore_test_${Date.now()}.kore`);

    try {
      // Write
      await kore.writeFile(tmpFile, block);
      assert(fs.existsSync(tmpFile));

      // Read
      const restored = await kore.readFile(tmpFile);

      assert.strictEqual(restored.numRows, 5);
      assert.strictEqual(restored.numColumns, 2);

      // Verify columns
      const numbersCol = restored.getColumn('numbers');
      assert(numbersCol !== undefined);
      assert.deepStrictEqual(numbersCol!.data, [1, 2, 3, 4, 5]);

      const namesCol = restored.getColumn('names');
      assert(namesCol !== undefined);
      assert.deepStrictEqual(namesCol!.data, ['a', 'b', 'c', 'd', 'e']);
    } finally {
      if (fs.existsSync(tmpFile)) {
        fs.unlinkSync(tmpFile);
      }
    }
  });

  it('should write and read float column', async () => {
    const block = kore.createDataBlock();
    block.addColumn('decimals', kore.DataType.F64, [1.1, 2.2, 3.3]);

    const tmpFile = path.join('/tmp', `kore_test_float_${Date.now()}.kore`);

    try {
      await kore.writeFile(tmpFile, block);
      const restored = await kore.readFile(tmpFile);

      const col = restored.getColumn('decimals');
      assert(col !== undefined);
      assert.strictEqual(col!.data.length, 3);
    } finally {
      if (fs.existsSync(tmpFile)) {
        fs.unlinkSync(tmpFile);
      }
    }
  });

  it('should write and read boolean column', async () => {
    const block = kore.createDataBlock();
    block.addColumn('flags', kore.DataType.BOOL, [true, false, true]);

    const tmpFile = path.join('/tmp', `kore_test_bool_${Date.now()}.kore`);

    try {
      await kore.writeFile(tmpFile, block);
      const restored = await kore.readFile(tmpFile);

      const col = restored.getColumn('flags');
      assert(col !== undefined);
      assert.deepStrictEqual(col!.data, [true, false, true]);
    } finally {
      if (fs.existsSync(tmpFile)) {
        fs.unlinkSync(tmpFile);
      }
    }
  });
});

// ─────────────────────────────────────────────────────────────────────────────
// COLUMN STATS (Phase 3)
// ─────────────────────────────────────────────────────────────────────────────

describe('ColumnStats', () => {
  it('should create column statistics', () => {
    const stats: kore.ColumnStats = {
      minValue: 1,
      maxValue: 100,
      nullCount: 0,
      cardinality: 50,
      crc32: 0xdeadbeef,
    };

    assert.strictEqual(stats.minValue, 1);
    assert.strictEqual(stats.maxValue, 100);
    assert.strictEqual(stats.nullCount, 0);
    assert.strictEqual(stats.cardinality, 50);
    assert.strictEqual(stats.crc32, 0xdeadbeef);
  });
});

// ─────────────────────────────────────────────────────────────────────────────
// VERSION CONTROL (Phase 3)
// ─────────────────────────────────────────────────────────────────────────────

describe('VersionControl', () => {
  it('should read at version (Phase 3: pending)', () => {
    // Phase 3: Implement once version snapshots integrated
  });
});

// ─────────────────────────────────────────────────────────────────────────────
// ENCRYPTION (Phase 3)
// ─────────────────────────────────────────────────────────────────────────────

describe('Encryption', () => {
  it('should encrypt and decrypt (Phase 3: pending)', () => {
    // Phase 3: Implement once crypto FFI exposed
  });
});

// ─────────────────────────────────────────────────────────────────────────────
// BLOOM FILTERS (Phase 3)
// ─────────────────────────────────────────────────────────────────────────────

describe('BloomFilters', () => {
  it('should retrieve Bloom filter (Phase 3: pending)', () => {
    // Phase 3: Implement once filter APIs exposed
  });
});

// Run tests if this is main module
if (require.main === module) {
  console.log('Run with: npx mocha kore_fileformat.test.ts');
}

File: kore-go/kore_fileformat_test.go
package kore

import (
	"testing"
)

// ─────────────────────────────────────────────────────────────────────────────
// DATA TYPES
// ─────────────────────────────────────────────────────────────────────────────

func TestDataTypeValues(t *testing.T) {
	tests := []struct {
		name     string
		dataType DataType
		expected uint8
	}{
		{"I64", I64, 1},
		{"F64", F64, 2},
		{"BOOL", BOOL, 3},
		{"STR", STR, 4},
		{"STR_DICT", STR_DICT, 5},
		{"ARRAY", ARRAY, 6},
		{"STRUCT", STRUCT, 7},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if uint8(tt.dataType) != tt.expected {
				t.Errorf("got %d, want %d", tt.dataType, tt.expected)
			}
		})
	}
}

func TestCompressionValues(t *testing.T) {
	tests := []struct {
		name        string
		compression Compression
		expected    uint8
	}{
		{"RAW", RAW, 0},
		{"RLE", RLE, 1},
		{"DELTA", DELTA, 2},
		{"DICT", DICT, 3},
		{"NAN_RAW", NAN_RAW, 4},
		{"DEFLATE", DEFLATE, 5},
		{"ZSTD", ZSTD, 6},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if uint8(tt.compression) != tt.expected {
				t.Errorf("got %d, want %d", tt.compression, tt.expected)
			}
		})
	}
}

func TestDataTypeString(t *testing.T) {
	tests := []struct {
		dataType DataType
		expected string
	}{
		{I64, "I64"},
		{F64, "F64"},
		{BOOL, "BOOL"},
		{STR, "STR"},
		{STR_DICT, "STR_DICT"},
		{ARRAY, "ARRAY"},
		{STRUCT, "STRUCT"},
	}

	for _, tt := range tests {
		t.Run(tt.expected, func(t *testing.T) {
			if tt.dataType.String() != tt.expected {
				t.Errorf("got %s, want %s", tt.dataType.String(), tt.expected)
			}
		})
	}
}

// ─────────────────────────────────────────────────────────────────────────────
// DATA BLOCK
// ─────────────────────────────────────────────────────────────────────────────

func TestNewDataBlock(t *testing.T) {
	db := NewDataBlock()

	if db == nil {
		t.Fatal("NewDataBlock returned nil")
	}
	if db.NumRows != 0 {
		t.Errorf("got NumRows %d, want 0", db.NumRows)
	}
	if db.NumColumns() != 0 {
		t.Errorf("got NumColumns %d, want 0", db.NumColumns())
	}
}

func TestAddColumn(t *testing.T) {
	db := NewDataBlock()

	// Add first column
	data := []int64{1, 2, 3, 4, 5}
	if err := db.AddColumn("numbers", I64, data); err != nil {
		t.Fatalf("AddColumn failed: %v", err)
	}

	if db.NumRows != 5 {
		t.Errorf("got NumRows %d, want 5", db.NumRows)
	}
	if db.NumColumns() != 1 {
		t.Errorf("got NumColumns %d, want 1", db.NumColumns())
	}

	// Add second column with same row count
	names := []string{"a", "b", "c", "d", "e"}
	if err := db.AddColumn("names", STR, names); err != nil {
		t.Fatalf("AddColumn failed: %v", err)
	}

	if db.NumColumns() != 2 {
		t.Errorf("got NumColumns %d, want 2", db.NumColumns())
	}
}

func TestAddColumnMismatchedRows(t *testing.T) {
	db := NewDataBlock()

	// Add first column with 5 rows
	if err := db.AddColumn("numbers", I64, []int64{1, 2, 3, 4, 5}); err != nil {
		t.Fatalf("AddColumn failed: %v", err)
	}

	// Try to add column with 3 rows (should fail)
	if err := db.AddColumn("names", STR, []string{"a", "b", "c"}); err == nil {
		t.Error("AddColumn should have failed with mismatched row count")
	}
}

func TestGetColumn(t *testing.T) {
	db := NewDataBlock()

	data := []int64{10, 20, 30}
	if err := db.AddColumn("test", I64, data); err != nil {
		t.Fatalf("AddColumn failed: %v", err)
	}

	col := db.GetColumn("test")
	if col == nil {
		t.Fatal("GetColumn returned nil")
	}

	if col.Name != "test" {
		t.Errorf("got Name %q, want test", col.Name)
	}
	if col.Type != I64 {
		t.Errorf("got Type %d, want %d", col.Type, I64)
	}
}

func TestGetColumnNotFound(t *testing.T) {
	db := NewDataBlock()
	db.AddColumn("test", I64, []int64{1, 2, 3})

	col := db.GetColumn("nonexistent")
	if col != nil {
		t.Errorf("GetColumn should return nil for nonexistent column, got %v", col)
	}
}

// ─────────────────────────────────────────────────────────────────────────────
// COLUMN STATS
// ─────────────────────────────────────────────────────────────────────────────

func TestColumnStats(t *testing.T) {
	minVal := 1.0
	maxVal := 100.0

	stats := &ColumnStats{
		MinValue:    &minVal,
		MaxValue:    &maxVal,
		NullCount:   0,
		Cardinality: 50,
		CRC32:       0xdeadbeef,
	}

	if *stats.MinValue != 1.0 {
		t.Errorf("got MinValue %f, want 1.0", *stats.MinValue)
	}
	if *stats.MaxValue != 100.0 {
		t.Errorf("got MaxValue %f, want 100.0", *stats.MaxValue)
	}
	if stats.NullCount != 0 {
		t.Errorf("got NullCount %d, want 0", stats.NullCount)
	}
}

// ─────────────────────────────────────────────────────────────────────────────
// CRC32
// ─────────────────────────────────────────────────────────────────────────────

func TestCRC32(t *testing.T) {
	// Test with simple data
	data := []byte{1, 2, 3, 4, 5}
	crc := CRC32(data)

	if crc == 0 {
		t.Error("CRC32 returned 0")
	}

	// CRC32 of same data should be consistent
	crc2 := CRC32(data)
	if crc != crc2 {
		t.Errorf("CRC32 not consistent: got %d and %d", crc, crc2)
	}
}

func TestCRC32Empty(t *testing.T) {
	crc := CRC32([]byte{})
	if crc != 0 {
		t.Errorf("CRC32 of empty slice should be 0, got %d", crc)
	}
}

// ─────────────────────────────────────────────────────────────────────────────
// VERSION
// ─────────────────────────────────────────────────────────────────────────────

func TestVersion(t *testing.T) {
	if Version != "2.0.0" {
		t.Errorf("got Version %s, want 2.0.0", Version)
	}
}
